src/Compilers/Core/Portable/CodeGen/DataStringHolder.cs

```diff
@@ -0,0 +1,424 @@
+        new ConcurrentDictionary<(ImmutableArray<byte> Data, ushort Alignment), (MappedField Bytes, StringField String)>(PrivateImplementationDetails.DataAndUShortEqualityComparer.Instance);
+
+    // synthesized methods
+    private ImmutableArray<Cci.IMethodDefinition> _orderedSynthesizedMethods;
+
+    public DataStringHolder(
+        CommonPEModuleBuilder moduleBuilder,
+        string dataHash,
+        Cci.ITypeReference systemObject,
+        Cci.ITypeReference systemString,
+        Cci.ICustomAttribute compilerGeneratedAttribute,
+        Cci.IMethodReference encodingUtf8,
+        Cci.IMethodReference encodingGetString,
+        PrivateImplementationDetails privateImplementationDetails)
+    {
+        _name = TypeNamePrefix + dataHash;
+        _moduleBuilder = moduleBuilder;
+        _systemObject = systemObject;
+        _systemString = systemString;
+        _compilerGeneratedAttribute = compilerGeneratedAttribute;
+        _encodingUtf8 = encodingUtf8;
+        _encodingGetString = encodingGetString;
+        _privateImplementationDetails = privateImplementationDetails;
+    }
+
+    public string Name => _name;
+
+    public string NamespaceName => string.Empty;
+
+    public bool IsPublic => false;
+
+    internal bool IsFrozen => _frozen != 0;
+
+    public override Cci.INamespaceTypeDefinition AsNamespaceTypeDefinition(EmitContext context) => this;
+
+    public override Cci.INamespaceTypeReference AsNamespaceTypeReference => this;
+
+    public Cci.IFieldReference CreateDataField(ImmutableArray<byte> data)
+    {
+        Cci.ITypeReference type = _privateImplementationDetails.GetOrAddProxyType(data.Length, alignment: 1);
+
+        Debug.Assert(_mappedFields.IsEmpty, "We need to generate unique field names if we are synthesizing more than one.");
+
+        (_, StringField stringField) = _mappedFields.GetOrAdd((data, Alignment: 1), key =>
+        {
+            var mappedField = new MappedField("f", this, type, data);
+            var stringField = new StringField(this, mappedField);
+            return (mappedField, stringField);
+        });
+
+        return stringField;
+    }
+
+    public override void Dispatch(Cci.MetadataVisitor visitor) => visitor.Visit(this);
+
+    public void Freeze(DiagnosticBag diagnostics)
+    {
+        var wasFrozen = Interlocked.Exchange(ref _frozen, 1);
+        if (wasFrozen != 0)
+        {
+            throw new InvalidOperationException();
+        }
+
+        var fieldsBuilder = ArrayBuilder<SynthesizedStaticField>.GetInstance(_mappedFields.Count * 2);
+
+        foreach (var (mappedField, stringField) in _mappedFields.Values)
+        {
+            fieldsBuilder.Add(mappedField);
+            fieldsBuilder.Add(stringField);
+        }
+
+        fieldsBuilder.Sort(PrivateImplementationDetails.FieldComparer.Instance);
+
+        _orderedSynthesizedFields = fieldsBuilder.ToImmutableAndFree();
```

Comment:
>_orderedSynthesizedFields = fieldsBuilder.ToImmutableAndFree(); [](http://example.com/codeflow?start=8&length=63)

It looks like we always synthesize two fields. It feels like all the code around ```fieldsBuilder``` is unnecessary complicated.